### PR TITLE
Distributed communication docs for Lite

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds docs for distributed communication with illustrations.

**Preview docs:**
1. Scroll down on the PR and look for "read the docs" in the build section.
2. Click on "Details". A preview of the docs build opens. 
3. Navigate to the Fabric docs and click on "Collectives" in the section "Advanced Topics"

image

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: docs/source-pytorch/fabric/advanced/collectives.rst [hunk 1]
```diff
@@ -4,45 +4,237 @@
...
 
 .. code-block:: python
 
-    fabric = Fabric()
+    from lightning.fabric import Fabric
 
-    # Transfer and concatenate tensors across processes
-    fabric.all_gather(...)
+    # Devices and num_nodes determine how many processes there are
+    fabric = Fabric(devices=2, num_nodes=3)
```
Comment (carmocca) [inline docs/source-pytorch/fabric/advanced/collectives.rst:35]: If #14078 lands then we wouldn't want to mention num_nodes like this, no?

File: docs/source-pytorch/fabric/advanced/collectives.rst [hunk 2]
```diff
@@ -4,45 +4,237 @@
...
         ...
 
-    # Wait for all processes to enter this call.
+
+.. _race conditions:
+
+Avoid race conditions
+=====================
+
+Access to the rank information helps you avoid *racing conditions* which could crash your script or lead to corrupted data.
```
Comment (carmocca) [inline docs/source-pytorch/fabric/advanced/collectives.rst:65]:
```suggestion
Access to the rank information helps you avoid *race conditions* which could crash your script or lead to corrupted data.
```

File: docs/source-pytorch/fabric/advanced/collectives.rst [hunk 3]
```diff
@@ -4,45 +4,237 @@
...
+
+    # Multi-node: download a dataset, the filesystem between nodes is NOT shared
+    if fabric.local_rank == 0:
+        download_dataset()
+
+----
+
+
+*******
+Barrier
```
Comment (carmocca) [inline docs/source-pytorch/fabric/advanced/collectives.rst:92]: Can you put these collective operations into it's own section called collectives?

And link PyTorch's docs: https://pytorch.org/docs/stable/distributed.html

This way this is ready for when our collectives are integrated.

File: docs/source-pytorch/fabric/advanced/collectives.rst [hunk 4]
```diff
@@ -4,45 +4,237 @@
...
+
+.. code-block:: python
+
+    fabric = Fabric(...)
+
+    # Gather the data from
+    result = fabric.all_gather(tensor)
+
+    # By default, you can't back-propagate through the all-gather operation
+    result = fabric.all_gather(tensor, sync_grads=False)
```
Comment (carmocca) [inline docs/source-pytorch/fabric/advanced/collectives.rst:201]: I wouldn't mention sync_grads here, as I believe we will remove it eventually in favor of using the `no_grad` context manager outside of the `all_gather`

**Review by carmocca — APPROVED** [review]
Love the graphics!

File: docs/source-pytorch/fabric/advanced/collectives.rst (hunk 3 above)
Comment (awaelchli) [reply docs/source-pytorch/fabric/advanced/collectives.rst:92]: I tried it, but it doesn't look so good because the titles get really small if they are grouped under yet another level. Actually, the whole page was already meant to be the section where we list the collectives.

File: docs/source-pytorch/fabric/advanced/collectives.rst (hunk 4 above)
Comment (awaelchli) [reply docs/source-pytorch/fabric/advanced/collectives.rst:201]: But that would be weird. I don't think torch would ever invert this behavior. It is like the reverse of what you want most of the time, this would be asking you to add `torch.no_grad()` most of the time, eventually becoming boilerplate code. I'm not sure what this idea is motivated by. Any pointers?

File: docs/source-pytorch/fabric/advanced/collectives.rst (hunk 1 above)
Comment (awaelchli) [reply docs/source-pytorch/fabric/advanced/collectives.rst:35]: Yes. In this case `world_size` would just be whatever is set externally on the cluster. For now it should be ok to show it like this, because it will take some work to complete https://github.com/Lightning-AI/lightning/issues/14078

**Review by justusschock — APPROVED** [review]
Love the graphics! Maybe I'd rename `fundamentals/distributed operation` then? As I thought this would be covered there as well. And link to the collectives docs from there?

File: docs/source-pytorch/fabric/advanced/distributed_communication.rst [hunk 5]
```diff
@@ -0,0 +1,240 @@
+:orphan:
+
+###########################################
+Communication between distributed processes
+###########################################
+
+With Fabric you can easily access information about a process or send data between processes with a standardized API and agnostic to the distributed strategy.
```
Comment (Borda) [inline docs/source-pytorch/fabric/advanced/distributed_communication.rst:7]:
```suggestion
With Fabric, you can easily access processes or send data between processes with a standardized API that is agnostic to the selected distributed strategy.
```

File: docs/source-pytorch/fabric/advanced/distributed_communication.rst [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+
+
+----
+
+
+*******************
+Rank and world size
+*******************
+
+The rank assigned to a process is a number in the range of *0, ..., world size - 1*, where *world size* is the total number of distributed processes.
```
Comment (Borda) [inline docs/source-pytorch/fabric/advanced/distributed_communication.rst:17]:
```suggestion
The rank assigned to a process is a zero-base index of *0, ..., world size - 1*, where *world size* is the total number of distributed processes.
```

File: docs/source-pytorch/fabric/advanced/distributed_communication.rst [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+Learn more about :doc:`launching distributed training <../fundamentals/launch>`.
+And here is how you access all rank and world size information:
+
+.. code-block:: python
+
+    # The total number of processes running across all devices and nodes
+    fabric.world_size  # 2 * 3 = 6
+
+    # The global index of the current process across all devices and nodes
+    fabric.global_rank
```
Comment (Borda) [inline docs/source-pytorch/fabric/advanced/distributed_communication.rst:47]:
```suggestion
    fabric.global_rank  # -> {0, 1, 2, 3, 4, 5}
```

File: docs/source-pytorch/fabric/advanced/distributed_communication.rst [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+    # The index of the current process among the processes running on the local node
+    fabric.local_rank
```
Comment (Borda) [inline docs/source-pytorch/fabric/advanced/distributed_communication.rst:50]:
```suggestion
    fabric.local_rank  # -> {0, 1, 2}
```

File: docs/source-pytorch/fabric/advanced/distributed_communication.rst [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+
+    # The index of the current node
+    fabric.node_rank
```
Comment (Borda) [inline docs/source-pytorch/fabric/advanced/distributed_communication.rst:53]:
```suggestion
    fabric.node_rank  # -> {0, 1}
```

File: docs/source-pytorch/fabric/advanced/collectives.rst (hunk 4 above)
Comment (carmocca) [reply docs/source-pytorch/fabric/advanced/collectives.rst]: We want to align our collectives API with PyTorch's and this is a difference that can become problematic. This boilerplate is managed by the PL loops, and in Fabric you will be able to easily turn on or off these context managers.

I don't mean to discuss this here. My suggestion is to not add docs that mention things that will likely go away

**Review by Borda — APPROVED** [review]

File: docs/source-pytorch/fabric/advanced/distributed_communication.rst (hunk 5 above)
Comment (justusschock) [reply docs/source-pytorch/fabric/advanced/distributed_communication.rst:7]: you can't actually access the process. So I prefer @awaelchli formulation!

File: docs/source-pytorch/fabric/advanced/distributed_communication.rst (hunk 6 above)
Comment (justusschock) [reply docs/source-pytorch/fabric/advanced/distributed_communication.rst:17]: `zero-based` index :)
